File: quant_proto_phase01/surface/lib/bridge.ml
(** Bridge: Call the Python Phase 0-4C compiler from OCaml.

    This module provides a subprocess-based bridge to the Python compiler.
    Terms are serialized to JSON, passed to bridge.py, and results parsed.
*)

(** Wire permutation result from Python *)
type wire_perm = {
  n : int;
  new_to_old : int list;
}

(** Compilation result *)
type compile_result =
  | CompileOk of wire_perm * int  (* perm, circuit_size *)
  | CompileError of string

(** Involution check result *)
type involution_result =
  | InvolutionOk of bool * wire_perm  (* is_invol, perm *)
  | InvolutionError of string

(** Path to bridge.py (relative to surface/) *)
let bridge_path = "bridge.py"

(** Path to Python interpreter *)
let python_path = "python3"

(** Path to project root (for activating venv) *)
let project_root = ref ""

(** Set the project root path *)
let set_project_root path = project_root := path

(** Convert a Rep.t to JSON type representation *)
let rec type_to_json = function
  | Rep.Var _ -> {|{"node": "Q"}|}  (* Variables become Q for now *)
  | Rep.Unit -> {|{"node": "Q"}|}   (* Unit becomes Q for now *)
  | Rep.Tensor (a, b) ->
    Printf.sprintf {|{"node": "Ten", "left": %s, "right": %s}|}
      (type_to_json a) (type_to_json b)
  | Rep.Plus (a, b) ->
    Printf.sprintf {|{"node": "Plus", "left": %s, "right": %s}|}
      (type_to_json a) (type_to_json b)

(** Term representation for JSON serialization *)
type term =
  (* Structural combinators *)
  | TId of Rep.t
  | TSeq of term * term
  | TTenTerm of term * term
  | TTwistTen of Rep.t * Rep.t
  | TAssocTenL of Rep.t * Rep.t * Rep.t
  | TAssocTenR of Rep.t * Rep.t * Rep.t
  | TTwistPlus of Rep.t * Rep.t
  | TAssocPlusL of Rep.t * Rep.t * Rep.t
  | TAssocPlusR of Rep.t * Rep.t * Rep.t
  (* Distributivity (unitary-level) *)
  | TDistL of Rep.t * Rep.t * Rep.t
  | TDistR of Rep.t * Rep.t * Rep.t
  (* Single-qubit gates *)
  | TH of int
  | TS of int
  | TSdg of int
  | TT of int
  | TTdg of int
  | TX of int
  | TY of int
  | TZ of int
  | TRx of float * int
  | TRy of float * int
  | TRz of float * int
  | TPhase of float * int
  (* Two-qubit gates *)
  | TCX of int * int
  | TCZ of int * int
  | TCRz of float * int * int
  (* Three-qubit gate *)
  | TCCX of int * int * int
  (* Controlled single-qubit gates for quantum case expressions *)
  | TCH of int * int
  | TCS of int * int
  | TCSdg of int * int

(** Convert a term to JSON *)
let rec term_to_json = function
  (* Structural combinators *)
  | TId ty ->
    Printf.sprintf {|{"node": "Id", "ty": %s}|} (type_to_json ty)
  | TSeq (f, g) ->
    Printf.sprintf {|{"node": "Seq", "f": %s, "g": %s}|}
      (term_to_json f) (term_to_json g)
  | TTenTerm (f, g) ->
    Printf.sprintf {|{"node": "TenTerm", "f": %s, "g": %s}|}
      (term_to_json f) (term_to_json g)
  | TTwistTen (a, b) ->
    Printf.sprintf {|{"node": "TwistTen", "a": %s, "b": %s}|}
      (type_to_json a) (type_to_json b)
  | TAssocTenL (a, b, c) ->
    Printf.sprintf {|{"node": "AssocTenL", "a": %s, "b": %s, "c": %s}|}
      (type_to_json a) (type_to_json b) (type_to_json c)
  | TAssocTenR (a, b, c) ->
    Printf.sprintf {|{"node": "AssocTenR", "a": %s, "b": %s, "c": %s}|}
      (type_to_json a) (type_to_json b) (type_to_json c)
  | TTwistPlus (a, b) ->
    Printf.sprintf {|{"node": "TwistPlus", "a": %s, "b": %s}|}
      (type_to_json a) (type_to_json b)
  | TAssocPlusL (a, b, c) ->
    Printf.sprintf {|{"node": "AssocPlusL", "a": %s, "b": %s, "c": %s}|}
      (type_to_json a) (type_to_json b) (type_to_json c)
  | TAssocPlusR (a, b, c) ->
    Printf.sprintf {|{"node": "AssocPlusR", "a": %s, "b": %s, "c": %s}|}
      (type_to_json a) (type_to_json b) (type_to_json c)
  (* Distributivity *)
  | TDistL (a, b, c) ->
    Printf.sprintf {|{"node": "DistL", "a": %s, "b": %s, "c": %s}|}
      (type_to_json a) (type_to_json b) (type_to_json c)
  | TDistR (a, b, c) ->
    Printf.sprintf {|{"node": "DistR", "a": %s, "b": %s, "c": %s}|}
      (type_to_json a) (type_to_json b) (type_to_json c)
  (* Single-qubit gates *)
  | TH i -> Printf.sprintf {|{"node": "H", "i": %d}|} i
  | TS i -> Printf.sprintf {|{"node": "S", "i": %d}|} i
  | TSdg i -> Printf.sprintf {|{"node": "Sdg", "i": %d}|} i
  | TT i -> Printf.sprintf {|{"node": "T", "i": %d}|} i
  | TTdg i -> Printf.sprintf {|{"node": "Tdg", "i": %d}|} i
  | TX i -> Printf.sprintf {|{"node": "X", "i": %d}|} i
  | TY i -> Printf.sprintf {|{"node": "Y", "i": %d}|} i
  | TZ i -> Printf.sprintf {|{"node": "Z", "i": %d}|} i
  | TRx (theta, i) -> Printf.sprintf {|{"node": "Rx", "theta": %f, "i": %d}|} theta i
  | TRy (theta, i) -> Printf.sprintf {|{"node": "Ry", "theta": %f, "i": %d}|} theta i
  | TRz (theta, i) -> Printf.sprintf {|{"node": "Rz", "theta": %f, "i": %d}|} theta i
  | TPhase (theta, i) -> Printf.sprintf {|{"node": "Phase", "theta": %f, "i": %d}|} theta i
  (* Two-qubit gates *)
  | TCX (i, j) -> Printf.sprintf {|{"node": "CX", "i": %d, "j": %d}|} i j
  | TCZ (i, j) -> Printf.sprintf {|{"node": "CZ", "i": %d, "j": %d}|} i j
  | TCRz (theta, i, j) -> Printf.sprintf {|{"node": "CRz", "theta": %f, "i": %d, "j": %d}|} theta i j
  (* Three-qubit gate *)
  | TCCX (i, j, k) -> Printf.sprintf {|{"node": "CCX", "i": %d, "j": %d, "k": %d}|} i j k
  (* Controlled single-qubit gates for quantum case expressions *)
  | TCH (i, j) -> Printf.sprintf {|{"node": "CH", "i": %d, "j": %d}|} i j
  | TCS (i, j) -> Printf.sprintf {|{"node": "CS", "i": %d, "j": %d}|} i j
  | TCSdg (i, j) -> Printf.sprintf {|{"node": "CSdg", "i": %d, "j": %d}|} i j

(** Simple JSON parsing helpers *)
let find_string key json =
  let pattern = Printf.sprintf {|"%s": "|} key in
  try
    let start = String.index json '"' in
    let json_from_key = String.sub json (String.index json (String.get pattern 0)) (String.length json - String.index json (String.get pattern 0)) in
    let _ = (start, json_from_key) in
    (* Simplified: look for "key": "value" or "key": value *)
    let re = Str.regexp (Printf.sprintf {|"%s": *"\([^"]*\)"|} key) in
    if Str.string_match re json 0 then
      Some (Str.matched_group 1 json)
    else
      let re2 = Str.regexp (Printf.sprintf {|"%s": *\([^,}]*\)|} key) in
      if Str.string_match re2 json 0 then
        Some (String.trim (Str.matched_group 1 json))
      else
        None
  with _ -> None

let find_bool key json =
  let re = Str.regexp (Printf.sprintf {|"%s": *\(true\|false\)|} key) in
  try
    let _ = Str.search_forward re json 0 in
    Some (Str.matched_group 1 json = "true")
  with Not_found -> None

let find_int key json =
  let re = Str.regexp (Printf.sprintf {|"%s": *\([0-9]+\)|} key) in
  try
    let _ = Str.search_forward re json 0 in
    Some (int_of_string (Str.matched_group 1 json))
  with Not_found -> None

let find_int_list key json =
  let re = Str.regexp (Printf.sprintf {|"%s": *\[\([0-9, ]*\)\]|} key) in
  try
    let _ = Str.search_forward re json 0 in
    let list_str = Str.matched_group 1 json in
    let parts = String.split_on_char ',' list_str in
    Some (List.map (fun s -> int_of_string (String.trim s)) parts)
  with _ -> None

(** Parse a wire_perm from JSON *)
let parse_perm json =
  match find_int "n" json, find_int_list "new_to_old" json with
  | Some n, Some new_to_old -> Some { n; new_to_old }
  | _ -> None

(** Call the Python bridge with a JSON request *)
let call_bridge request_json =
  let bridge_script = Filename.concat !project_root "surface/bridge.py" in
  let venv_python = Filename.concat !project_root "../venv/bin/python" in

  (* Write request to temp file *)
  let tmp_in = Filename.temp_file "qpl_bridge_" ".json" in
  let tmp_out = Filename.temp_file "qpl_bridge_" ".out" in

  let oc = open_out tmp_in in
  output_string oc request_json;
  close_out oc;

  (* Run python with temp file I/O *)
  let cmd = Printf.sprintf "%s %s < %s > %s 2>&1"
    venv_python bridge_script tmp_in tmp_out in
  let _ = Sys.command cmd in

  (* Read response *)
  let ic = open_in tmp_out in
  let len = in_channel_length ic in
  let output = really_input_string ic len in
  close_in ic;

  (* Cleanup *)
  Sys.remove tmp_in;
  Sys.remove tmp_out;

  String.trim output

(** Compile a term and return the wire permutation *)
let compile term =
  let term_json = term_to_json term in
  let request = Printf.sprintf {|{"type": "compile", "term": %s}|} term_json in

  let response = call_bridge request in

  match find_bool "success" response with
  | Some true ->
    (match parse_perm response, find_int "circuit_size" response with
     | Some perm, Some size -> CompileOk (perm, size)
     | _ -> CompileError "Failed to parse response")
  | Some false ->
    (match find_string "error" response with
     | Some err -> CompileError err
     | None -> CompileError "Unknown error")
  | None ->
    CompileError ("Invalid response: " ^ response)

(** Check if a term compiles to an involutive permutation *)
let check_involution term =
  let term_json = term_to_json term in
  let request = Printf.sprintf {|{"type": "check_involution", "term": %s}|} term_json in

  let response = call_bridge request in

  match find_bool "success" response with
  | Some true ->
    (match find_bool "is_involution" response, parse_perm response with
     | Some is_invol, Some perm -> InvolutionOk (is_invol, perm)
     | _ -> InvolutionError "Failed to parse response")
  | Some false ->
    (match find_string "error" response with
     | Some err -> InvolutionError err
     | None -> InvolutionError "Unknown error")
  | None ->
    InvolutionError ("Invalid response: " ^ response)

(** Helper: create a TwistPlus term for a 2-constructor datatype swap *)
let twist_plus_for_bool () =
  TTwistPlus (Rep.var 0, Rep.var 1)

File: quant_proto_phase01/surface/bridge.py
#!/usr/bin/env python3
"""Bridge script for OCaml surface language integration.

This script provides a JSON-based interface for the OCaml surface compiler
to call the Phase 0-4C Python compiler.

Usage:
    echo '{"type": "compile", "term": {...}}' | python bridge.py
    echo '{"type": "check_involution", "term": {...}}' | python bridge.py

Input JSON format for terms:
    {"node": "TwistPlus", "a": {"node": "Q"}, "b": {"node": "Q"}}
    {"node": "Id", "ty": {"node": "Plus", "left": {"node": "Q"}, "right": {"node": "Q"}}}
"""

import json
import sys
from pathlib import Path

# Add src to path
src_path = Path(__file__).parent.parent / "src"
sys.path.insert(0, str(src_path))

from lang.types import Q, Ten, Plus, Unit, Ty, width
from lang.terms import (
    Term, Id, Seq, TenTerm,
    TwistTen, AssocTenL, AssocTenR,
    TwistPlus, AssocPlusL, AssocPlusR,
    DistL, DistR,
    H, S, Sdg, T, Tdg, X, Y, Z,
    Rx, Ry, Rz, Phase,
    CX, CZ, CRz, CCX,
    # Controlled single-qubit gates
    CH, CS, CSdg,
    # Higher-order constructs (GOI apply)
    FunVar, Lam, Apply,
    # Exponentials of structural involutions
    ExpSwap, ExpInvolution,
    # Bifunctorial action on sums
    PlusMap,
    # Case combinator
    Case,
)
from core.perm import WirePerm, identity, compose
from compile.to_pytket import compile
from compile.goi import (
    GOIArtifact, GateAtom, LoopSpec,
    apply_perm, goi_seq, make_unitary_value, conjugate_unitary,
)


def parse_type(j: dict) -> Ty:
    """Parse a JSON type representation into a Ty."""
    node = j["node"]
    if node == "Q":
        return Q()
    elif node == "Unit" or node == "I":
        return Unit()
    elif node == "Ten":
        return Ten(parse_type(j["left"]), parse_type(j["right"]))
    elif node == "Plus":
        return Plus(parse_type(j["left"]), parse_type(j["right"]))
    else:
        raise ValueError(f"Unknown type node: {node}")


def _max_wire_index(j: dict) -> int:
    """Find the maximum wire index used in a JSON term."""
    node = j.get("node", "")
    max_idx = -1

    # Check for wire indices in gates
    if "i" in j:
        max_idx = max(max_idx, j["i"])
    if "j" in j:
        max_idx = max(max_idx, j["j"])
    if "k" in j:
        max_idx = max(max_idx, j["k"])

    # Recurse into subterms
    if "f" in j:
        max_idx = max(max_idx, _max_wire_index(j["f"]))
    if "g" in j:
        max_idx = max(max_idx, _max_wire_index(j["g"]))
    if "body" in j:
        max_idx = max(max_idx, _max_wire_index(j["body"]))
    if "arg" in j:
        max_idx = max(max_idx, _max_wire_index(j["arg"]))
    if "left" in j:
        max_idx = max(max_idx, _max_wire_index(j["left"]))
    if "right" in j:
        max_idx = max(max_idx, _max_wire_index(j["right"]))

    return max_idx


def parse_general_gate(j: dict, ty_total: Ty) -> Term:
    """Parse a general gate with arbitrary controls.

    JSON format: {"node": "Gate", "name": "H", "targets": [1], "controls": [0, 2]}

    For nested quantum case expressions, gates can have multiple controls
    (one for each nesting level).
    """
    from lang.terms import (
        Id, Seq,
        H, S, Sdg, T, Tdg, X, Y, Z,
        CH, CS, CSdg,
    )

    name = j["name"]
    targets = j["targets"]
    controls = j.get("controls", [])

    if len(targets) != 1:
        raise ValueError(f"General gate currently only supports single-target gates, got {targets}")

    target = targets[0]

    # No controls: just the basic gate
    if len(controls) == 0:
        if name == "H":
            return H(target, ty_total)
        elif name == "S":
            return S(target, ty_total)
        elif name == "Sdg":
            return Sdg(target, ty_total)
        elif name == "T":
            return T(target, ty_total)
        elif name == "Tdg":
            return Tdg(target, ty_total)
        elif name == "X":
            return X(target, ty_total)
        elif name == "Y":
            return Y(target, ty_total)
        elif name == "Z":
            return Z(target, ty_total)
        else:
            raise ValueError(f"Unknown gate name: {name}")

    # Single control: use existing controlled gates
    elif len(controls) == 1:
        ctrl = controls[0]
        if name == "H":
            return CH(ctrl, target, ty_total)
        elif name == "S":
            return CS(ctrl, target, ty_total)
        elif name == "Sdg":
            return CSdg(ctrl, target, ty_total)
        else:
            raise ValueError(f"No single-controlled version of gate: {name}")

    # Multiple controls: decompose using the first control + recursion
    # C^n[G] = C[C^{n-1}[G]] implemented as nested controlled gates
    else:
        # For multi-controlled gates, we decompose:
        # CC...C[G] with controls [c0, c1, ..., cn] on target t
        # becomes a sequence that uses ancilla or direct decomposition
        #
        # For now, use the simple (but gate-expensive) decomposition:
        # Apply controlled version with first control, where the "base gate"
        # is itself controlled by the remaining controls
        #
        # This is handled by the to_pytket compiler which supports multi-controlled ops
        raise ValueError(
            f"Multi-controlled gate ({len(controls)} controls) not yet supported in bridge. "
            f"Gate: {name}, controls: {controls}, target: {target}"
        )


def _build_ty_total(n_qubits: int) -> Ty:
    """Build a tensor type with n_qubits qubits: Q ⊗ Q ⊗ ... ⊗ Q."""
    if n_qubits <= 0:
        return Q()
    if n_qubits == 1:
        return Q()

    # Build right-associated: Q ⊗ (Q ⊗ (Q ⊗ ...))
    result = Q()
    for _ in range(n_qubits - 1):
        result = Ten(Q(), result)
    return result


def parse_term(j: dict, ty_total: Ty = None) -> Term:
    """Parse a JSON term representation into a Term.

    ty_total: The total type context for gates. If None, inferred from max wire index.
    """
    # Infer ty_total from max wire index if not provided
    if ty_total is None:
        max_idx = _max_wire_index(j)
        n_qubits = max(2, max_idx + 1)  # At least 2 qubits
        ty_total = _build_ty_total(n_qubits)

    node = j["node"]

    # Structural combinators
    if node == "Id":
        return Id(parse_type(j["ty"]))

    elif node == "Seq":
        return Seq(parse_term(j["f"], ty_total), parse_term(j["g"], ty_total))

    elif node == "TenTerm":
        return TenTerm(parse_term(j["f"], ty_total), parse_term(j["g"], ty_total))

    elif node == "TwistTen":
        return TwistTen(parse_type(j["a"]), parse_type(j["b"]))

    elif node == "AssocTenL":
        return AssocTenL(parse_type(j["a"]), parse_type(j["b"]), parse_type(j["c"]))

    elif node == "AssocTenR":
        return AssocTenR(parse_type(j["a"]), parse_type(j["b"]), parse_type(j["c"]))

    elif node == "TwistPlus":
        return TwistPlus(parse_type(j["a"]), parse_type(j["b"]))

    elif node == "AssocPlusL":
        return AssocPlusL(parse_type(j["a"]), parse_type(j["b"]), parse_type(j["c"]))

    elif node == "AssocPlusR":
        return AssocPlusR(parse_type(j["a"]), parse_type(j["b"]), parse_type(j["c"]))

    # Distributivity
    elif node == "DistL":
        return DistL(parse_type(j["a"]), parse_type(j["b"]), parse_type(j["c"]))

    elif node == "DistR":
        return DistR(parse_type(j["a"]), parse_type(j["b"]), parse_type(j["c"]))

    # Single-qubit gates
    elif node == "H":
        return H(j["i"], ty_total)

    elif node == "S":
        return S(j["i"], ty_total)

    elif node == "Sdg":
        return Sdg(j["i"], ty_total)

    elif node == "T":
        return T(j["i"], ty_total)

    elif node == "Tdg":
        return Tdg(j["i"], ty_total)

    elif node == "X":
        return X(j["i"], ty_total)

    elif node == "Y":
        return Y(j["i"], ty_total)

    elif node == "Z":
        return Z(j["i"], ty_total)

    elif node == "Rx":
        return Rx(j["theta"], j["i"], ty_total)

    elif node == "Ry":
        return Ry(j["theta"], j["i"], ty_total)

    elif node == "Rz":
        return Rz(j["theta"], j["i"], ty_total)

    elif node == "Phase":
        return Phase(j["theta"], j["i"], ty_total)

    # Two-qubit gates
    elif node == "CX":
        return CX(j["i"], j["j"], ty_total)

    elif node == "CZ":
        return CZ(j["i"], j["j"], ty_total)

    elif node == "CRz":
        return CRz(j["theta"], j["i"], j["j"], ty_total)

    # Three-qubit gate
    elif node == "CCX":
        return CCX(j["i"], j["j"], j["k"], ty_total)

    # Controlled single-qubit gates (for quantum case expressions)
    elif node == "CH":
        return CH(j["i"], j["j"], ty_total)

    elif node == "CS":
        return CS(j["i"], j["j"], ty_total)

    elif node == "CSdg":
        return CSdg(j["i"], j["j"], ty_total)

    # General multi-controlled gate (for nested cases)
    elif node == "Gate":
        return parse_general_gate(j, ty_total)

    # Higher-order constructs (GOI apply)
    elif node == "FunVar":
        return FunVar(j["name"], parse_type(j["dom"]), parse_type(j["cod"]))

    elif node == "Lam":
        body = parse_term(j["body"], ty_total)
        return Lam(j["name"], parse_type(j["dom"]), parse_type(j["cod"]), body)

    elif node == "Apply":
        return Apply(parse_term(j["f"], ty_total), parse_term(j["arg"], ty_total))

    # Bifunctorial action on sums (⊕-Map)
    elif node == "PlusMap":
        ty_left = parse_type(j["ty_left"])
        ty_right = parse_type(j["ty_right"])
        # Branches operate on payload types, not full sum type
        left = parse_term(j["left"], ty_left)
        right = parse_term(j["right"], ty_right)
        return PlusMap(ty_left, ty_right, left, right)

    # Pattern-matching case expression from OCaml Linear DSL
    # CaseExpr = Seq(scrut, Case(ty_left, ty_right, left, right))
    elif node == "CaseExpr":
        ty_left = parse_type(j["ty_left"])
        ty_right = parse_type(j["ty_right"])
        scrut = parse_term(j["scrut"], ty_total)
        # Branches operate on payload types, not full sum type
        left = parse_term(j["left"], ty_left)
        right = parse_term(j["right"], ty_right)
        # Compose: first scrutinee, then case combinator
        case_combinator = Case(ty_left, ty_right, left, right)
        return Seq(scrut, case_combinator)

    # Exponentials of structural involutions
    elif node == "ExpSwap":
        return ExpSwap(j["theta"], j["i"], j["j"], ty_total)

    elif node == "ExpInvolution":
        body = parse_term(j["body"], ty_total)
        return ExpInvolution(j["theta"], body, ty_total)

    else:
        raise ValueError(f"Unknown term node: {node}")


def perm_to_json(p: WirePerm) -> dict:
    """Convert a WirePerm to JSON."""
    return {
        "n": p.n,
        "new_to_old": p.new_to_old
    }


def is_involution(p: WirePerm) -> bool:
    """Check if p ∘ p = identity."""
    p_squared = compose(p, p)
    id_perm = identity(p.n)
    return p_squared == id_perm


def handle_compile(request: dict) -> dict:
    """Handle a compile request."""
    try:
        term = parse_term(request["term"])
        result = compile(term, materialize=False)

        return {
            "success": True,
            "perm": perm_to_json(result.perm),
            "circuit_size": result.circuit.n_gates
        }
    except Exception as e:
        return {
            "success": False,
            "error": str(e)
        }


def handle_check_involution(request: dict) -> dict:
    """Handle an involution check request.

    With tagged layout model, structural sum operations (like TwistPlus)
    emit X gates for tag flips. These are still considered structural
    because X·X = I, so the tag flips cancel when composed.

    We check:
    1. Only X gates allowed (tag flips from sum operations)
    2. Permutation is involutive (p ∘ p = identity)
    """
    try:
        term = parse_term(request["term"])
        result = compile(term, materialize=False)

        # Check that only X gates are present (tag flips are OK)
        for cmd in result.circuit.get_commands():
            if cmd.op.type.name != 'X':
                return {
                    "success": False,
                    "error": f"Term is not structural (contains {cmd.op.type.name} gate)"
                }

        # Check involution
        is_invol = is_involution(result.perm)

        return {
            "success": True,
            "is_involution": is_invol,
            "perm": perm_to_json(result.perm)
        }
    except Exception as e:
        return {
            "success": False,
            "error": str(e)
        }


def main():
    """Main entry point."""
    # Read JSON from stdin
    input_data = sys.stdin.read()

    try:
        request = json.loads(input_data)
    except json.JSONDecodeError as e:
        print(json.dumps({"success": False, "error": f"Invalid JSON: {e}"}))
        sys.exit(1)

    req_type = request.get("type", "compile")

    if req_type == "compile":
        response = handle_compile(request)
    elif req_type == "check_involution":
        response = handle_check_involution(request)
    else:
        response = {"success": False, "error": f"Unknown request type: {req_type}"}

    print(json.dumps(response))


if __name__ == "__main__":
    main()
